internal/abci/abci.go

```diff
@@ -685,14 +681,71 @@ func (a *AbciApp) prepareMempoolTxns(txs [][]byte, maxBytes int, log *log.Logger
 
 		if bytes.Equal(tx.Sender, proposerAddr) {
 			proposerNonce = tx.Body.Nonce + 1
-		}
+			propTxs = append(propTxs, tx)
+		} else {
+			// Append the transaction to the final list.
+			if _, ok := otherTxns[string(tx.Sender)]; !ok {
+				otherTxns[string(tx.Sender)] = make([]*indexedTxn, 0)
+				senders = append(senders, string(tx.Sender))
 
-		finalTxns = append(finalTxns, txs[tx.is])
+			}
+			otherTxns[string(tx.Sender)] = append(otherTxns[string(tx.Sender)], tx)
+		}
 		nonces = append(nonces, tx.Body.Nonce)
 		i++
 	}
 
-	return finalTxns, proposerNonce
+	// truncate based on our max block size since we'll have to set
+	// ConsensusParams.Block.MaxBytes to -1 so that we get ALL transactions even
+	// if it goes beyond max_tx_bytes.  See:
+	// https://github.com/cometbft/cometbft/pull/1003
+	// https://docs.cometbft.com/v0.38/spec/abci/abci++_methods#prepareproposal
+	// https://github.com/cometbft/cometbft/issues/980
+
+	// Enforce block size limits
+	// Txs order: MempoolProposerTxns, ProposerInjectedTxns, MempoolTxns
+	var finalTxs [][]byte
+
+	for _, tx := range propTxs {
+		txBts := txs[tx.is]
+		txSize := int64(len(txBts))
+		if maxTxBytes < txSize {
+			break
+		}
+		maxTxBytes -= txSize
+		finalTxs = append(finalTxs, txBts)
+	}
+
+	proposerTxs, err := a.txApp.ProposerTxs(context.Background(), proposerNonce, maxTxBytes, proposerAddr)
+	if err != nil {
+		log.Error("failed to get proposer transactions", zap.Error(err))
+	}
+
+	for _, tx := range proposerTxs {
+		txSize := int64(len(tx))
+		if maxTxBytes < txSize {
+			break
+		}
+		maxTxBytes -= txSize
+		finalTxs = append(finalTxs, tx)
+	}
+
+	for _, sender := range senders {
```

Comment:
A comment on the `senders` loop, and above when building the slice would be helpful.

Reply:
It's not obvious why the mempool transactions should be grouped by sender, or if that's really desirable.

For instance, we could do without `senders` slices, and simplify the bottom mempool loop:

```
	for sender, senderTxs := range senders {
		for _, tx := range senderTxs {
```

That is still grouped, but random order of the groups.

Reply:
I think I added it, so that its testable. But probably we don't have to group them by sender

Reply:
Updated, so that the order is still fifo, except for proposer txns.